peep.c

```diff
@@ -3934,6 +3934,11 @@ Perl_rpeep(pTHX_ OP *o)
                   * child (PADSV), and gets to it via op_other rather
                   * than op_next. Don't try to optimize this. */
                  && (lval != rhs)
+                 /* For efficiency, pp_padsv_store() doesn't push its
+                  * result onto the stack. For the relatively rare case of
+                  * the $lex assignment not in void context, we just do it
+                  * the old slow way. */
+                 && OP_GIMME(o,0) == G_VOID
```

Comment:
I wondered how rare, so I measured it, first, add some code so we can distinguish the cases:
```
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -307,6 +307,8 @@ PP(pp_and)
  *    (PL_op->op_private & OPpASSIGN_BACKWARDS) {or,and,dor}assign
 */
 
+unsigned void_stores;
+
 PP(pp_padsv_store)
 {
     OP * const op = PL_op;
@@ -336,6 +338,8 @@ PP(pp_padsv_store)
         );
     SvSetMagicSV(targ, val);
 
+    if (GIMME_V == G_VOID)
+        ++void_stores;
     rpp_replace_1_1_NN(targ);
     return NORMAL;
 }
```
build and test with `-fprofile-arcs -ftest-coverage`, and review the result:
```
382305756:  341:    if (GIMME_V == G_VOID)
320134726:  342:        ++void_stores;
```
which is 83.7% void context, or 1 in 6 are non-void context.